Discover existing snapshots on disk

ashish-amarnath opened the pull request:
Signed-off-by: Ashish Amarnath <ashisham@vmware.com>

<!--  Thanks for sending a pull request!  Here are some tips for you:

1. If this is your first time, please read our contributor guidelines: https://git.k8s.io/community/contributors/guide#your-first-contribution and developer guide https://git.k8s.io/community/contributors/devel/development.md#development-guide
2. Follow the instructions for writing a release note: https://git.k8s.io/community/contributors/guide/release-notes.md
3. If the PR is unfinished, see how to mark it: https://git.k8s.io/community/contributors/guide/pull-requests.md#marking-unfinished-pull-requests
-->

**What type of PR is this?**
/kind feature

**What this PR does / why we need it**:
This change lets this driver discover existing snapshots on disk and therefore improves testability of the scenario of "Importing an existing volume snapshot with Kubernetes"

**Which issue(s) this PR fixes**:
<!--
*Automatically closes linked issue when PR is merged.
Usage: `Fixes #<issue number>`, or `Fixes (paste link of issue)`.
_If PR is about `failing-tests or flakes`, please post the related issues/tests in a comment and do not use `Fixes`_*
-->
Fixes #

**Special notes for your reviewer**:

**Does this PR introduce a user-facing change?**:
<!--
If no, just write "NONE" in the release-note block below.
If yes, a release note is required:
Enter your extended release note in the block below. If the PR requires additional action from users switching to the new release, include the string "action required".
-->
```release-note
Adds ability to discover and import existing on-disk tarballs as snapshots.
```

ashish-amarnath commented:
This was verified on a 1.17 cluster.
Logs from testing:
```
I0228 23:26:55.219236       1 controllerserver.go:554] Enabling controller service capability: CREATE_DELETE_SNAPSHOT
I0228 23:26:55.219248       1 controllerserver.go:554] Enabling controller service capability: LIST_SNAPSHOTS
I0228 23:26:55.219263       1 controllerserver.go:554] Enabling controller service capability: CLONE_VOLUME
I0228 23:26:55.219275       1 controllerserver.go:554] Enabling controller service capability: EXPAND_VOLUME
I0228 23:26:55.219316       1 hostpath.go:129] discovering existing snapshots in /csi-data-dir
I0228 23:26:55.219369       1 hostpath.go:147] snapshot: /csi-data-dir/deadfeed.tgz, id: deadfeed
I0228 23:26:55.219582       1 server.go:100] Listening for connections on address: &net.UnixAddr{Name:"//csi/csi.sock", Net:"unix"}
I0228
```

Inspecting storage objects

```
$ kc get volumesnapshotcontent
NAME                                READYTOUSE   RESTORESIZE   DELETIONPOLICY   DRIVER                VOLUMESNAPSHOTCLASS      VOLUMESNAPSHOT         AGE
snapcontent-hand-crafted-deadfeed   true         0             Delete           hostpath.csi.k8s.io   csi-hostpath-snapclass   velero-csi-pvc-9jwhg   31m

$ kcn csi-app get volumesnapshot
NAME                   READYTOUSE   SOURCEPVC   SOURCESNAPSHOTCONTENT               RESTORESIZE   SNAPSHOTCLASS   SNAPSHOTCONTENT                     CREATIONTIME   AGE
velero-csi-pvc-9jwhg   true                     snapcontent-hand-crafted-deadfeed   0                             snapcontent-hand-crafted-deadfeed   50y            30m

$ kcn csi-app get pv
NAME                                       CAPACITY   ACCESS MODES   RECLAIM POLICY   STATUS   CLAIM             STORAGECLASS      REASON   AGE
pvc-063555a0-dc11-48f7-be58-b6d12c9da482   1Gi        RWO            Delete           Bound    csi-app/csi-pvc   csi-hostpath-sc            30m

$ kcn csi-app get pvc
NAME      STATUS   VOLUME                                     CAPACITY   ACCESS MODES   STORAGECLASS      AGE
csi-pvc   Bound    pvc-063555a0-dc11-48f7-be58-b6d12c9da482   1Gi        RWO            csi-hostpath-sc   30m

```

xing-yang commented:
/ok-to-test

xing-yang commented:
This is very useful to test pre-provisioned snapshots.

Can you also add steps on how you manually create those snapshots?
This is "/kind feature".   Need a release note.

ashish-amarnath commented:
@xing-yang  I've added `CHANGELOG-1.4.md` w/ release notes for this PR.
Wasn't sure where I should add the steps to create the snapshots.

xing-yang commented:
@ashish-amarnath For release note, you only need:
Adds ability to discover and import existing on-disk tarballs as snapshots. 

The PR info will be taken care when this driver is released.
It's fine to add the steps here in this PR.  We can decide where to document this later.

xing-yang commented:
> On doing this, the snapshot controller along with the hostpath plugin driver will provision a pv and load it with the data from the snapshot, datafeed.tgz

It should be the provisioner not the snapshot controller that provisions the PV.

xing-yang commented on pkg/hostpath/hostpath.go:
```diff
@@ -127,12 +128,42 @@ func NewHostPathDriver(driverName, nodeID, endpoint string, ephemeral bool, maxV
+	glog.V(4).Infof("discovering existing snapshots in %s", dataRoot)
+	snapshots := []string{}
+	err := filepath.Walk(dataRoot, func(path string, info os.FileInfo, err error) error {
+		glog.V(6).Infof("path: %s", path)
```
V(4) is fine.

xing-yang commented:
> 6. Now to create PVs using this snapsthot

s/snapsthot/snapshot

You are only creating PV that uses the snapshot as data source, but not creating the source PV/PVC for this snapshot, right?

ashish-amarnath commented:
@xigang 
> It should be the provisioner not the snapshot controller that provisions the PV.
Thanks for the correction 😄 I've updated the comment.

> You are only creating PV that uses the snapshot as data source, but not creating the source PV/PVC for this snapshot, right?

I am creating `volumesnapshotcontent`, `volumesnapshot` and then the `pvc` using the volumesnapshot as the dataSource.

I've also updated the comment w/ this clarification.

xing-yang commented:
Can you rework this based on this change?  https://github.com/kubernetes-csi/csi-driver-host-path/pull/162

xing-yang commented:
Can you modify your steps to add a step to rename *.tgz files to *.snap?

xing-yang commented:
/lgtm
/approve

xing-yang commented:
/assign @msau42 
/assign @jsafrane

msau42 commented on pkg/hostpath/hostpath.go:
```diff
@@ -127,12 +131,39 @@ func NewHostPathDriver(driverName, nodeID, endpoint string, ephemeral bool, maxV
+	err := filepath.Walk(dataRoot, func(path string, info os.FileInfo, err error) error {
+		glog.V(4).Infof("path: %s", path)
+		if filepath.Ext(path) == snapshotExt {
+			pathSplit := strings.Split(path, ".")[0]
```
Can a filename having multiple "." in it?

Can you add a comment to what is the expected syntax for `path`?

ashish-amarnath replied:
done

ashish-amarnath replied:
yeah files like `foo.tar.gz` have multiple `.` in its name

xing-yang replied:
We can probably support foo.tar.snap too?

ashish-amarnath replied:
Yes, based on the [golang library](https://golang.org/pkg/path/filepath/#Ext) `filepath.Ext`, `foo.tar.snap` also has a an extension `.snap`

xing-yang replied:
Can you add this to the example?
s/e.g. deadfeed.snap/ e.g. foo.snap, foo.bar.snap

ashish-amarnath replied:
Done!

jsafrane commented on pkg/hostpath/hostpath.go:
```diff
@@ -127,12 +131,41 @@ func NewHostPathDriver(driverName, nodeID, endpoint string, ephemeral bool, maxV
+		if filepath.Ext(path) == snapshotExt {
+			// Files with .snap extension are volumesnapshot files.
+			// e.g. deadfeed.snap
+			pathSplit := strings.Split(path, ".")[0]
```
this removes everything after the first dot (`foo.bar.snap` -> `foo`), while the result should be `foo.bar`.

ashish-amarnath replied:
Thanks for catching that!
I've fixed that bug and also added unit tests for it.

xing-yang commented on pkg/hostpath/hostpath_test.go:
```diff
@@ -0,0 +1,58 @@
+			expectedIsSnapshot: true,
+			expectedSnapshotID: "baz.tar",
+		},
+	}
```
Can you add a test for "foo/bar/baz.tar.snap.snap"?

ashish-amarnath replied:
added

xing-yang commented on pkg/hostpath/hostpath.go:
```diff
@@ -127,12 +132,46 @@ func NewHostPathDriver(driverName, nodeID, endpoint string, ephemeral bool, maxV
+	// Files with .snap extension are volumesnapshot files.
+	// e.g. foo.snap, foo.bar.snap
+	if filepath.Ext(file) == snapshotExt {
+		snapshotID := strings.Split(file, snapshotExt)[0]
```
If file is "baz.tar.snap.snap", it will return "baz.tar".  It should return "baz.tar.snap".  Only the last .snap should be removed.

ashish-amarnath replied:
That's a bug. Thank you for catching that!
It is now fixed.

xing-yang commented:
/lgtm

msau42 commented on pkg/hostpath/hostpath.go:
```diff
@@ -127,12 +132,45 @@ func NewHostPathDriver(driverName, nodeID, endpoint string, ephemeral bool, maxV
+
+func discoverExistingSnapshots() {
+	glog.V(4).Infof("discovering existing snapshots in %s", dataRoot)
+	err := filepath.Walk(dataRoot, func(p string, i os.FileInfo, e error) error {
```
Do we need to do a walk here? Are snapshots stored in a nested structure, or a single directory?

ashish-amarnath replied:
We don't need to `walk` here, the snapshots are files under the `dataRoot`.
I'll update the code to do read the dir.

msau42 commented on pkg/hostpath/hostpath_test.go:
```diff
@@ -0,0 +1,63 @@
+	}
+
+	for _, tc := range testCases {
+		actualIsSnapshot, actualSnapshotID := getSnapshotID(tc.inputPath)
```
I would wrap each test case in `t.Run`

ashish-amarnath replied:
done

msau42 commented on pkg/hostpath/hostpath_test.go:
```diff
@@ -0,0 +1,63 @@
+	for _, tc := range testCases {
+		actualIsSnapshot, actualSnapshotID := getSnapshotID(tc.inputPath)
+		if actualIsSnapshot != tc.expectedIsSnapshot {
+			t.Fatalf("getSnapshotID failed, unexpected result for path %s, Want: %t, Got: %t", tc.inputPath, tc.expectedIsSnapshot, actualIsSnapshot)
```
These should be `Errorf` instead of `Fatalf` so that other test cases will still run even if one fails.

xing-yang commented:
/lgtm

msau42 commented:
/approve
